docs/markdown/using-snowchain/validating-and-relaying/relayer-flow.md

```diff
@@ -0,0 +1,127 @@
+---
+layout: default
+title: Relayer Flow
+parent: Validating and Relaying
+nav_order: 2
+---
+
+# Relayer Flow
+
+## Polkadot to Ethereum
+The relayer can be started in 2 different modes for relaying from Polkadot to Ethereum.
+
+## MMR-Epochs-Only mode
+This mode runs the relayer in a mode where it ensures that the Ethereum contract receives at least 1 MMR proof per epoch. It only interacts with the Relay Chain Light Client, not the Parachain Light Client.
```

Comment:
When you say 1 per epoch @musnit, do you mean Polkadot epoch or Ethereum epoch? The diagram below also shows that you're watching Eth blocks, not Polkadot blocks. I prob missing something. How do Ethereum blocks and MMR roots relate?

Reply:
@Rizziepit Polkadot epoch. Epochs mentioned anywhere in this context refer to polkadot epochs, as eth epochs are not relevant.

MMR R
And correct, we do watch eth blocks too, basically:
 - MMR Roots are basically like new Polkadot blocks
 - MMR Roots are relayed to Ethereum and ethereum stores the latest MMR Root that has been relayed after the 2-step interactive relay process completes.
 - The relayer needs to keep track of the most recent Polkadot Epoch that Ethereum is aware of. This is latestRelayedEpoch.

We watch Polkadot to detect new MMR Roots that have been **generated**, so that we can relay them to Ethereum.
We watch Ethereum to detect new MMR Roots that have been **relayed** (eg: by someone else, or by the relayer in a different mode) so that if they contain a new epoch, we can update our awareness of Ethereum to this new epoch.

Can discuss in more detail on next call if needed.

Reply:
The MMR-Epochs-Only is basically intended only to ensure that the light client contract never skips an epoch, nothing else. In a normal scenario, where there is actual message flow through the bridge and at least one message per epoch, the Incentivized-Relay mode should pickup that message and relay the respective required MMR Root for it, and so the process running in MMR-Epochs-Only node effectively does nothing/never relays. It's only important in cases where there are no messages at all for more than 1 epoch.

I've also updated the New Incentivized Channel Commitment Processing Queue and the preview link to include some changes that reflect how the Incentivized-Relay mode should pickup and relay required MMR Roots - please review.

Reply:
> We watch Ethereum to detect new MMR Roots that have been relayed (eg: by someone else, or by the relayer in a different mode) so that if they contain a new epoch, we can update our awareness of Ethereum to this new epoch.

In the Eth -> Polkadot direction, we don't continuously keep track of the latest Eth block that Polkadot is aware of. We only retrieve this info when the relayer starts, and then forward each new block as we become aware of it.

So my question is, how does continously keeping track of the latest Polkadot epoch that Ethereum is aware of change this logic? At best we can save some network bandwidth by skipping an MMR root send if we know it's already out of date... but it's prob simpler to send it regardless, as soon as we become aware of it, and hope for the best. Why won't this strategy work?

Reply:
I think because in MMR-Epochs-Only mode, the expected default is that another relayer running in Incentivized-Relay mode will be moving the light client forward new blocks and so the MMR-Epochs-Only mode won't really ever need to forward blocks in many circumstances. It could do this anyway without keeping track, as you say, but consider the following:
 - Need to handle failure properly, detecting why and then skipping over onto the next one as in normal circumstances failure due to someone else relaying first is likely
 - The relay is a 2-transaction process, not 1-transaction, so again may require watching Ethereum to handle failure anyway, depending on when the failure occurs, though I haven't thought about this too much
 - It's not just about saving network bandwidth, its also about saving ethereum gas costs, which will be significant because the cost for the 2-transaction process is high even in the case of failure/reversion.

Having said that, your idea for simple/greedy strategy sounds fine to implement for MVP presuming it can handle failure and continue correctly, and then can optimize for saving gas by watching after.